feat(button): add "featured" variant

dancormier opened the pull request:
[STACKS-680](https://stackoverflow.atlassian.net/browse/STACKS-680)

## TODO

> [!NOTE]
> I've opened this PR up for review but there's a couple of outstanding items that need to be resolved:

- [x] Verify name "featured"
- [ ] Write description in docs

---

This PR adds a "featured" (purple) variant to the button component.

<details>
<summary>Screenshots</summary>

## Light

![image](https://github.com/user-attachments/assets/3387df5f-b0cc-4ce4-9081-1c49a06fc780)

## Dark

![image](https://github.com/user-attachments/assets/299cef03-ebcc-4fa6-9ca3-92a34d4ea0a9)

## Light high contrast

![image](https://github.com/user-attachments/assets/d5847be4-d2a6-478d-b864-9a2e528d2717)

## Dark high contrast

![image](https://github.com/user-attachments/assets/2141d4f0-7ec2-4ac1-8f56-d25d87391f95)

</details>

## Testing

1. Go to `/product/components/buttons/#featured`
2. Verify the newly added _featured_ buttons look and act as expected

## Notable changes

### High contrast filled background color

~~To meet accessibility requirements, the featured filled variant background color has been set to `--purple-500` (normally `--purple-400`).~~

We've modified the `purple-400` light HC mode color instead. See https://github.com/StackExchange/Stacks/pull/1857#pullrequestreview-2466079719

### Borders on filled, selected buttons

This PR includes a minor fix to button border colors when hovered and filled and/or selected. Previously, the value of `--bu-bc` (`currentColor`) was applied and now the border color applied is equal to `--bu-bg`, ensuring the border and button background are unified.

### Button group visual regression test images

There are six visual regression tests (all on `s-btn-group-highcontrast-*` in Chromium runs) that have been updated. They are visually identical to me but there could be extremely slight differences due to https://github.com/StackExchange/Stacks/pull/1857/commits/838e25ad90ab4b68e52b54c2e89e2b8b43bf4054.

dancormier committed:
feat(button): add "featured" variant

https://stackoverflow.atlassian.net/browse/STACKS-680

dancormier committed:
Modify high contrast filled bg color

dancormier committed:
Fix button border color quirks

When filled/selected buttons were hovered, the currentColor would be used for the border color. This commit ensures the correct border color is used on filled/selected buttons

dancormier committed:
Add featured buttons to docs

dancormier committed:
Modify test skipped ids to exclude appropriate featured buttons

dancormier committed:
Add visual test images

dancormier committed:
Update s-btn-group images

dancormier committed:
Revert "Update s-btn-group images"

This reverts commit 0907245e98ee3263de290ad7d87971a82c15f236.

dancormier committed:
Minor muted selected border color fix

dancormier commented on docs/product/components/buttons.html:
```diff
@@ -119,6 +120,61 @@
+
+    {% header "h3", "Featured" %}
+    <!-- TODO add featured button description -->
+    <p class="stacks-copy">Featured buttons are…</p>
```
@CGuindon Here's where the description for the "featured" buttons would go. Feel free to make a commit or pass a description along to me and I'll plug it in.

CGuindon replied:
I can't push to Stacks so here you go — let me know what you think. I looked at the other buttons and the color purple description in the color fundamentals:

> Featured buttons are a secondary button style, used to visually draw attention to something new or temporary, usually as part of onboarding or to announce a new feature. These should be used sparingly, and permanent placements should be avoided

dancormier commented on docs/_data/buttons.json:
```diff
@@ -35,6 +35,26 @@
     }
+  ]
+  ,
+  "featured": [
```
I went with `featured` for this variant name. 

I figured it's best to avoid coupling naming with visual that could change (`purple`) and `new` felt too restrictive since I could see this button being used just to grab some attention generally. `featured` felt like the best fit but I'm glad to change it if you feel differently @CGuindon.

giamir commented on lib/components/button/button.a11y.test.ts:
```diff
@@ -14,8 +14,8 @@ describe("button", () => {
             /s-btn-highcontrast-light-(?!.*(filled|danger|github|facebook|sm|xs)).*?badge/,
-            // matches tests with a badge in highcontrast-light modes, are muted and/or outlined, and are sm or xs
-            /s-btn-highcontrast-light-(?:muted-outlined-|muted-|outlined-)?(?:sm|xs).*?badge/,
+            // matches tests with a badge in highcontrast-light modes, are muted or featured and/or outlined, and are sm or xs
```
Out of curiosity what's the contrast ratio for the failing tests? 
By glancing at the UI it doesn't seem we are a lot off from the 7:1 target.

dancormier replied:
Yeah, the featured variant seems to land at 6.61 in the failing cases. It's from the button badge opacity being set to `.8` in high contrast mode.

_Personally_, I'd be fine to bump that to `.85` or `1`, but we'd either need to add an exception for the `featured` variant for this or change the button badge opacity for all buttons in HC mode. I'd prefer the latter but I'd want to get approval from design before making either change. Thoughts?

giamir replied:
I also think the latter is the better option. Then let's consider this out of scope for this PR and maybe let's have a chat with design about it at some point. Thanks

giamir commented on docs/_data/buttons.json:
```diff
@@ -85,6 +105,16 @@
           "class": "s-btn__danger",
           "class2": "s-btn__filled"
         },
+        {
```
We are missing the `Featured, Clear` combination

giamir commented on docs/_data/buttons.json:
```diff
@@ -127,6 +157,16 @@
           "class": "s-btn__danger",
           "class2": "s-btn__filled"
         },
+        {
```
We are missing the `Featured, Clear` combination

giamir commented on docs/_data/buttons.json:
```diff
@@ -169,6 +209,16 @@
           "class": "s-btn__danger",
           "class2": "s-btn__filled"
         },
+        {
```
We are missing the `Featured, Clear` combination

giamir reviewed (commented):
Overall looks good, there is only a minor docs correction to do I left in comment.

Nice to see the PPCP abstraction making our lives easier here. 
Well done @dancormier ❤️

dancormier committed:
Minor docs fixes

Addresses https://github.com/StackExchange/Stacks/pull/1857#discussion_r1856719220 and similar

dancormier committed:
Merge branch 'develop' into STACKS-680/button-add-featured-variant

dancormier committed:
Update new test images to .ico extension

CGuindon reviewed (commented):
Ok so the only thing I don't love is that in HC mode, the filled default state is the same color as the filled selected state (no difference). 

Because we had to bump the background color from 400 to 500 for HC filled, and because 500 and 600 are the same in HC, we end up with no difference there.

For Base and Danger we have a modified red-400 and blue-400 just for HC mode. Those colors were modified from Light/Dark so that they meet the 7:1 ratio (they're pretty close but not quite otherwise). And so for those variants we're able to keep the 400 stop as the background color for filled default and therefore have a color difference with the selected state.

So....

I'm proposing we modify HC modes purple 400 color to meet a 7:1 ratio with white.
Change the selected state back to using purple-400 but with this new color (in dark mode the purple 400 already passes 7:1 ratio.

- HC Light purple-400 would become #3B41BA ([Figma where I played with this](https://www.figma.com/design/X2txHfaagLc3lMSwnjBZnT/Colors%3A-Accessible-Color-Palette-v2?node-id=2170-14817&t=flVejKXMnu6uG3XU-1))
- HC Dark purple-400 would become  NO CHANGE NEEDED? since it already passes the 7:1 ratio?

---

<img width="827" alt="Screenshot 2024-11-27 at 3 08 26 PM" src="https://github.com/user-attachments/assets/0f4172d0-e1f1-49fb-8ae2-0febf348cdd5">

---

<img width="736" alt="Screenshot 2024-11-27 at 3 08 47 PM" src="https://github.com/user-attachments/assets/982d7ca6-39fd-448a-9223-15fcc6d51782">

dancormier committed:
Add featured variant description to docs

Sourced from here https://github.com/StackExchange/Stacks/pull/1857#discussion_r1861314934

dancormier committed:
Update purple-400 HC value

dancormier committed:
Remove featured HC filled bg override

dancormier committed:
Update visual regression test images

dancormier committed:
Merge branch 'develop' into STACKS-680/button-add-featured-variant
